### PR TITLE
[NFC][SYCL] Simplify the error checking of arrays by only visiting 1x.

### PR DESCRIPTION
**Pull request by erichkeane**
The current visitor requires that we go through our diagnostics checkers
once per element.  This is expensive, so this patch makes it an opt-out
as to whether we should visit each element, or only the 1st.

### CODE REVIEW

**Comment by erichkeane** [comment]
Please forgive there being two commits, the first is covered in this review: https://github.com/intel/llvm/pull/2345

**Comment by erichkeane** [comment]
I've now merged this with the dependent patch, so the 'diff' is now correct.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -863,17 +873,35 @@ class KernelObjVisitor {
     QualType ET = CAT->getElementType();
     int64_t ElemCount = CAT->getSize().getSExtValue();
     std::initializer_list<int>{(handlers.enterArray(), 0)...};
-    for (int64_t Count = 0; Count < ElemCount; Count++) {
-      VisitElement(nullptr, FD, ET, handlers...);
+
+    if (ElemCount > 0) {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:877]: Isn't it always true?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:877]: Nope, not necessarily apparently: https://godbolt.org/z/eox5ME

Its rare, but possible.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:877]: Is it zero-sized array? Didn't we prohibit them for SYCL?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:877]: Looks like you're right!  Thanks!

**Review by Fznamznon — APPROVED** [review]
LGTM, thanks

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -863,17 +873,35 @@ class KernelObjVisitor {
     QualType ET = CAT->getElementType();
     int64_t ElemCount = CAT->getSize().getSExtValue();
     std::initializer_list<int>{(handlers.enterArray(), 0)...};
-    for (int64_t Count = 0; Count < ElemCount; Count++) {
-      VisitElement(nullptr, FD, ET, handlers...);
+
+    assert(ElemCount > 0 && "SYCL prohibits 0 sized arrays");
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:877]: We prohibit size-zero arrays.  If the user attempts to use a size-zero array in a kernel, they get a diagnostic.

**Pull request merged**